src/opt/horizontal_fusion_impl.cpp

```diff
@@ -0,0 +1,839 @@
+#include "horizontal_fusion_impl.hpp"
+#include <migraphx/iterator_for.hpp>
+
+namespace migraphx {
+inline namespace MIGRAPHX_INLINE_NS {
+
+static unsigned opcode_bits  = 16;
+static unsigned hash_id_bits = 16;
+static unsigned filter_bits  = 8;
+static unsigned kernel_bits  = 8;
+
+// Register a single operation.
+// 1st arg: operation name. 2nd arg: encoding function.
+void horizontal_fusion_impl::register_op(const std::string& name, encoder func, int flag)
+{
+    op_registry[name] = std::move(func);
+    op_flag[name]     = flag;
+}
+
+// Register operations.
+void horizontal_fusion_impl::register_all()
+{
+    register_op("gpu::convolution", encode_conv_common, 1);
+    register_op("gpu::conv_bias_relu", encode_conv_common, 1);
+    register_op("hip::add_relu", encode_common, 0);
+    register_op("convolution", encode_conv_common, 1);
+    register_op("add", encode_common, 0);
+    register_op("relu", encode_common, 0);
+}
+
+static unsigned opcode_shift_count() { return ((sizeof(key_type) * 8) - opcode_bits); }
+
+static unsigned hash_id_shift_count() { return (opcode_shift_count() - hash_id_bits); }
+
+static unsigned filter_shift_count() { return (hash_id_shift_count() - filter_bits); }
+
+static unsigned kernel_shift_count() { return (filter_shift_count() - kernel_bits); }
+
+// Encode common fields.
+//
+// |----- 16 bits -----|----- 16 bits -------|----- 32 bits -----|
+// |      opcode       | 1st operand hash id |
+encode_info encode_common(instruction_ref ins, ins2_val& instr2_value, unsigned opcode)
+{
+    if(opcode >= (static_cast<unsigned>(1) << opcode_bits))
+        return encode_info(0, false);
+    key_type encode     = (static_cast<key_type>(opcode) << opcode_shift_count());
+    instruction_ref op1 = ins->inputs().front();
+    if(instr2_value.find(op1) == instr2_value.end())
+        return encode_info(0, false);
+    hash_value_ptr op1_val = instr2_value[op1];
+
+    if((op1_val == nullptr) || (op1_val->id >= (static_cast<unsigned>(1) << hash_id_bits)))
+        return encode_info(0, false);
+    encode |= (static_cast<key_type>(op1_val->id) << hash_id_shift_count());
+    encode_info info(encode, true);
+    info.add_input(op1_val);
+    return info;
+}
+
+// Encode common fields in convolution:
+//
+// |----- 16 bits -----|----- 16 bits -------|----- 8 bits -----|----- 8 bits-----|----- 16 bits
+// -----| |     opcode        | 1st operand hash id |   filter size    |  kernel size    |
+// 0x0000        |
```

Comment:
This should be written as bit fields rather than comments(as comments can become stale):

```cpp
struct conv_encoding
{
    unsigned opcode: 16;
    unsigned operand: 16;
    unsigned filter_size: 8;
    unsigned kernel_size: 8;
};
```